onnxruntime/core/graph/function.cc

```diff
@@ -7,6 +7,126 @@
 #include "core/graph/model.h"
 #include "onnx/shape_inference/implementation.h"
 
+namespace ONNX_NAMESPACE {
+// Infer shape for functions. This also supports 
+// nested model local functions.
+// TODO: Add this to onnx instead of adding it here.
+void InferShapeForFunctionNode(
+    InferenceContext& ctx,
+    const FunctionProto& func,
+    const std::unordered_map<std::string, int>& func_opset_imports,
+    const ShapeInferenceOptions& options,
+    const ISchemaRegistry* schema_registry,
+    const std::unordered_map<std::string, const FunctionProto*>& in_model_functions) {
+  GraphProto g;
+  // Get a temporary tensor-shape map
+  const auto num_func_inputs = func.input_size();
+  std::unordered_map<std::string, TypeProto*> valueTypesByName;
+  std::vector<TypeProto> types_cache(func.input_size());
```

Comment:
nit: some of the variable names don't conform to the coding standards

Reply:
If this is going to end up in ONNX it's probably fine to follow that convention, however things are inconsistent currently so pick one or the other.